src/concurrent.rs

```diff
@@ -0,0 +1,170 @@
+    /// that this happened because of us closing the file descriptor.  Unfortunately... macOS
+    /// Mojave doesn't cooperate and yields very strange error codes (like EISDIR) pretty frequently
+    /// (but not deterministically).  May it be an APFS bug?
+    closed: Arc<Mutex<bool>>,
+}
+
+impl ShareableFile {
+    /// Constructs a new `ShareableFile` from an open file and takes ownership of it.
+    pub fn from(file: fs::File) -> ShareableFile {
+        use std::os::unix::io::IntoRawFd;
+        ShareableFile {
+            fd: file.into_raw_fd(),
+            owned: true,
+            closed: Arc::from(Mutex::from(false)),
+        }
+    }
+
+    /// Returns an unowned view of the file.
+    ///
+    /// Users of this file must accept that the file can be closed at any time by the owner.
+    pub fn clone_unowned(&mut self) -> ShareableFile {
+        ShareableFile {
+            fd: self.fd,
+            owned: false,
+            closed: self.closed.clone(),
+        }
+    }
+}
+
+impl Drop for ShareableFile {
+    fn drop(&mut self) -> () {
+        if self.owned {
+            debug!("Closing ShareableFile with fd {}", self.fd);
+
+            // We must mark the file as closed before attempting to close it because, when we do
+            // the actual close, any threads blocked on a read will get unlocked immediately and
+            // we won't have a chance to update the status.  It doesn't matter if the close succeeds
+            // or not though, because we cannot do anything useful in the latter case anyway.
+            *self.closed.lock().unwrap() = true;
+
+            match unistd::close(self.fd) {
+                Ok(_) => (),
+                Err(e) => warn!("Failed to close fd {}: {}", self.fd, e)
+            }
+        }
+    }
+}
+
+impl io::Read for ShareableFile {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match unistd::read(self.fd, buf) {
+            Ok(n) => Ok(n),
+            Err(nix::Error::Sys(errno)) => {
+                if *self.closed.lock().unwrap() {
+                    Ok(0)  // Simulate EOF due to close in another thread.
+                } else {
+                    Err(io::Error::from_raw_os_error(errno as i32))
+                }
+            },
+            _ => panic!("Did not expect to get an error without an errno from a nix call"),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use nix::sys;
+    use std::io::{Read, Write};
+    use std::path::Path;
+    use std::thread;
+    use super::*;
+    use tempfile;
+
+    #[test]
+    fn test_shareable_file_clones_share_descriptor_and_only_one_owns() {
+        let dir = tempfile::tempdir().unwrap();
+        let path = dir.path().join("file");
+        {
+            let mut file = fs::File::create(&path).unwrap();
+            use std::io::Write;
+            write!(file, "ABCDEFG").unwrap();
+        }
+
+        fn read_one_byte<R: Read>(input: &mut R) -> u8 {
```

Comment:
Why dynamic dispatch over monomorphization?

Reply:
Also, shouldn't it be `&mut dyn Read`?

Reply:
You can put `&mut impl Read` if you want it to be explicitly monomorphic. I am just suggesting these because I find them a bit easier to read then the explicit constraints.

Reply:
> Also, shouldn't it be &mut dyn Read?

Right. I forgot about that change.

Reply:
Done.